crypto/x509/x_pubkey.c

```diff
@@ -472,3 +472,32 @@ ASN1_BIT_STRING *X509_get0_pubkey_bitstr(const X509 *x)
         return NULL;
     return x->cert_info.key->public_key;
 }
+
+/*
+ * Returns 0 for equal, -1 for a < b or error on a, 1 for a > b or error on b,
+ * -2 if the public keys are different, -3 if the key types are different, and
+ * -4 if the operation is not supported
+ */
+int X509_PUBKEY_cmp(X509_PUBKEY *a, X509_PUBKEY *b)
+{
+    X509_ALGOR *algA, *algB;
+    EVP_PKEY *pA, *pB;
+    int res;
+
+    if (a == b)
+        return 0;
+    if (a == NULL || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algA, a)
+            || algA == NULL)
+        return -1;
+    if (b == NULL || !X509_PUBKEY_get0_param(NULL, NULL, NULL, &algB, b)
+            || algB == NULL)
+        return 1;
+    if ((res = X509_ALGOR_cmp(algA, algB)) != 0)
+        return res;
+    if ((pA = X509_PUBKEY_get0(a)) == NULL)
+        return -1;
+    if ((pB = X509_PUBKEY_get0(b)) == NULL)
+        return 1;
+    res = EVP_PKEY_cmp(pA, pB);
+    return res == 1 ? 0 : res - 2;
```

Comment:
No this can't be right, the return values of EVP_PKEY_cmp
are :
1  = equel
0  = non-equal
-1 = key type mismatch
-2 = error

so no way how to return 1 for a > b here.
Anyway this returning +/- 1 based on numeric difference of private key data
looks more or less like a oracle.

Reply:
It is actually public key data. But anyway, I agree that there seems to be no point in other than equal/non-equal/error results.

Reply:
So I've changed the semantics from order comparison to equality, and renamed the function accordingly to `X509_PUBKEY_eq()`.